Source/Core/Monomorphization.cs

```diff
@@ -1293,11 +1297,20 @@ class MonomorphizationVisitor : VarDeclOnceStandardVisitor
     private HashSet<TypeCtorDecl> visitedTypeCtorDecls;
     private HashSet<Function> visitedFunctions;
     private Dictionary<Procedure, Implementation> procToImpl;
+    public readonly HashSet<Function> originalFunctions;
+    public readonly HashSet<Constant> originalConstants;
+    // Note that original axioms refer to axioms of the original program which might have been updated in-place during monomorphization.
+    public readonly Dictionary<Axiom, HashSet<Axiom>> originalAxiomToSplitAxioms;
+    public readonly Dictionary<Axiom, Axiom> originalAxiomToUninstantiatedCopies;
```

Comment:
How about this:
We add something like the following to `Absy.Axiom`
```
    public Axiom OriginalAxiom;
    public Axiom OriginalAxiomUninstantiatedCopy;
```
and update them during monomorphization. This is in line with instantiated `DeclWithFormals` [in Absy.cs](https://github.com/boogie-org/boogie/blob/0b1e6843c94480df8e05790f43ae4182e52e3204/Source/Core/AST/Absy.cs#L1680-L1682).

We could also add to `Absy.Axiom`:
```
    public ISet<Function> Functions;
    public ISet<Function> Constants;
```
which could return body functions & symbols for an axiom. Then we would no longer need the additional dictionaries. The function and constant collector could also be moved out somewhere, maybe to `Absy.cs`.

Reply:
I implemented the version I proposed above, but I am not sure if it ended up being any better. In the new implementation members of `Axiom` need to be initialized in the right places during monomorphization, as opposed to adding entries to the dictionaries. The dictionaries are no longer needed, but similar collections need to be computed again from `splitAxioms` inside `UpdateFunctionDependencies`. Maybe additional members in `Axiom`, which are only used during monomorphization, are also undesirable. Overall I am tempted to just change the name for now. This can perhaps be considered again when adding `used_by`.

Reply:
When I made the suggestion, I had something very simple in mind. You have the following lines of code in your PR:

```
var originalAxiomFc = new FunctionAndConstantCollector();
var uninstantiatedOriginalAxiom = monomorphizationVisitor.originalAxiomToUninstantiatedCopy[originalAxiom];
originalAxiomFc.Visit(uninstantiatedOriginalAxiom);
```

If I understand this correctly, you retrieving the cloned axiom (prior to in-place updates) just so you can call the `originalFc.Visit` method. You could call this method up front and stash originalAxiomFc in the dictionary instead. Then you can look up the collected functions and constants later.

Reply:
That makes sense, and axioms don't need to be cloned anymore. I just pushed a [commit](https://github.com/boogie-org/boogie/pull/767/commits/7b28f0e4ad1a73dfacdf59f68f0719db8caea47d) doing this, thanks for the suggestion!